### DISCUSSION TITLE
Add parameters to back-end url

### DISCUSSION THREAD

**Discussion by radlik01** [apiman/apiman#1441] [Help] [answered]
Is it possible to add some parameters or replace some part of url before sending it to back-end ?
I need to send some confidential data to back-end in url (get REST API) but I cannot use this parameters from public internet.
The confidential data are in token or in headers (i put some data fro token to header before sending request to back-end ).

public call:
https://server/apiman-gateway/portal_pacient/vaccination_calendar/1.0/cs/api/CalendarEvent/

change to back-end
https://server/apiman-gateway/portal_pacient/vaccination_calendar/1.0/cs/api/CalendarEvent/{patientId}

What is the best way to do it ?

Thank you

Jan

**Comment by volkflo** [comment]
Hi,

the simplest solution would be to use one of our policies.

- https://www.apiman.io/apiman-docs/user-guide/2.1.0.Final/gateway/policies.html#_url_rewriting_policy
  - This would allow you to rewrite the URL
- https://www.apiman.io/apiman-docs/user-guide/2.1.0.Final/gateway/policies.html#_simple_header_policy
  - If the backend allows Headers this would help to send the data as a header.
  
I think the only problem is, that these are all fixed values. 
Your `patientId`seems to be a dynamic value?

**Reply by radlik01** [reply]
Yes, the problem is that I need to work with dynamic values that comes from header or token (claims)

I have no problem to write custom policy or other plugin module.

Jan

**Reply by msavy** [reply]
Custom policy might be easiest and most compact, but you can extract token fields from the JWT using the Keycloak OAuth2 policy (should work with any JWT that isn't weird).

Interesting use-case.

**Reply by radlik01** [reply]
I will share the sources. But what type of Interface I should use. Because if I extend AbstractMappedPolicy the doApply has request marked as final  ApiRequest request.

What is the best point to inject my solution ?

Jan

**Reply by msavy** [reply]
You need to mutate the existing request fields (in-place) as Apiman uses this to perform its activities. For example, you may want to mutate the `destination` to make it hit a different destination path.

You likely want to `getDestination` do your business logic on the retrieved address, then `setDestination` with your new endpoint.

**Reply by radlik01** [reply]
Thank you. If setDestination is enough than it is easy to do it.

I will share the plugin and sources after success implementation

Jan

**Reply by msavy** [reply]
Cool, should be simple. Check out https://github.com/apiman/apiman-plugins for additional examples.

**Reply by volkflo** [reply]
Maybe it would be a cool addition to an existing policy:
Something that forwards headers or cliam values as query parameters instead of only pass them along.

**Reply by radlik01** [reply]
I will do something like that
`    String origDestination = request.getDestination();
    String newDestination = request.getDestination();

    for (Map.Entry<String, String> consumer: request.getHeaders().getEntries()) {
      if (sLog.isDebugEnabled()) {
        sLog.debugf("doApply header key=value %s=%s", consumer.getKey(), consumer.getValue());
      }
      newDestination = newDestination.replaceAll("[{" + consumer.getKey() + "}]", consumer.getValue());
    }

    request.setDestination(newDestination);
    if (sLog.isDebugEnabled()) {
      sLog.debugf("doApply after change destination %s %s", origDestination, newDestination);
    }
    if (successStatus.getValue()) {
      chain.doApply(request);
    }`

**Reply by radlik01** [reply]
Yes existing Simple Header policy will be the best

**Reply by msavy** [reply]
Something like that -- although you'd want to select the specific headers you're targeting (or extract from JWT), because there are loads of other HTTP headers which will not be relevant to your purposes 👍.

**Reply by radlik01** [reply]
Yes. I could put to config that only headers starting with X- should be processing

I have already deploy the plugin to test env and it seams that is working like a charm.

Thank you

**Comment by msavy** [comment, marked as answer]
Answer was to write a custom policy plugin. Please refer to the [comments here](https://github.com/apiman/apiman/discussions/1441#discussioncomment-1158058).
